File: SemanticSearch/PHP/Web/Plain/src/Storage/Adapters/Pinecone.php
<?php
require_once __DIR__ . '/../VectorAdapter.php';

class Pinecone implements VectorStoreAdapter {
    private ?object $index = null;

    public function connect(): void {
        if ($this->index !== null) return;
        $pinecone = new \Pinecone\Pinecone(getenv('PINECONE_API_KEY') ?: '');
        $indexName = getenv('PINECONE_INDEX') ?: 'documents';
        $this->index = $pinecone->index($indexName);
    }

    public function addDocuments(array $documents, array $embeddings, array $metadata): void {
        $this->connect();
        $vectors = [];
        foreach ($documents as $i => $doc) {
            $vectors[] = [
                'id' => "doc_$i",
                'values' => $embeddings[$i],
                'metadata' => array_merge($metadata[$i], ['text' => $doc]),
            ];
        }
        $this->index->upsert($vectors);
    }

    public function search(array $queryEmbedding, int $nResults = 5): array {
        $this->connect();
        $results = $this->index->query($queryEmbedding, $nResults, true);
        return array_map(fn($match) => [
            'document' => $match['metadata']['text'] ?? '',
            'metadata' => array_diff_key($match['metadata'], ['text' => true]),
            'distance' => $match['score'] ?? null,
        ], $results['matches'] ?? []);
    }

    public function deleteCollection(string $collectionName): void {}

    public function listCollections(): array {
        try {
            $pinecone = new \Pinecone\Pinecone(getenv('PINECONE_API_KEY') ?: '');
            return array_map(fn($i) => $i->name, $pinecone->listIndexes());
        } catch (Exception $e) {
            return [];
        }
    }

    public function close(): void {
        $this->index = null;
    }
}

File: SemanticSearch/Python/Web/Flask/src/storage/adapters/pinecone.py
import os
from storage.vector_adapter import VectorStoreAdapter


class Pinecone(VectorStoreAdapter):
    def __init__(self):
        self._index = None
        self.connect()

    def connect(self):
        if self._index is not None:
            return
        import pinecone
        api_key = os.getenv("PINECONE_API_KEY", "")
        environment = os.getenv("PINECONE_ENVIRONMENT", "us-west1-gcp")
        pinecone.init(api_key=api_key, environment=environment)
        index_name = os.getenv("PINECONE_INDEX", "documents")
        if index_name not in pinecone.list_indexes():
            pinecone.create_index(index_name, dimension=int(os.getenv("VECTOR_DIMENSION", "1536")))
        self._index = pinecone.Index(index_name)

    def add_documents(self, documents, embeddings, metadata):
        vectors = []
        for i, (doc, emb, meta) in enumerate(zip(documents, embeddings, metadata)):
            vectors.append({
                "id": f"doc_{i}",
                "values": emb,
                "metadata": meta | {"text": doc},
            })
        self._index.upsert(vectors=vectors)

    def search(self, query_embedding, n_results=5):
        if self._index is None:
            return []
        results = self._index.query(vector=query_embedding, top_k=n_results, include_metadata=True)
        return [
            {"document": match["metadata"].get("text", ""), "metadata": {k: v for k, v in match["metadata"].items() if k != "text"}, "distance": match["score"]}
            for match in results.get("matches", [])
        ]

    def delete_collection(self, collection_name):
        pass

    def list_collections(self):
        try:
            import pinecone
            return pinecone.list_indexes()
        except Exception:
            return []

    def close(self):
        self._index = None